File: src/nexus/core/object_store.py
"""ObjectStoreABC -- the Object Store pillar of the Four Storage Pillars.

Kernel ``file_operations`` contract (Linux analogue: ``struct file_operations``).
Exception-based errors, consistent with MetastoreABC / RecordStoreABC / CacheStoreABC.

Design:
    - ABC (not Protocol) -- concrete defaults for streaming, batch, capability flags
    - 6 abstract methods: write_content, read_content, delete_content,
      get_content_size, mkdir, rmdir
    - WriteResult returned from write operations (content_id + size)
    - Callers get raw types, errors are exceptions
"""

from __future__ import annotations

from abc import ABC, abstractmethod
from collections.abc import Iterator
from typing import TYPE_CHECKING

from nexus.contracts.types import WriteResult

if TYPE_CHECKING:
    from nexus.contracts.types import OperationContext

# Re-export WriteResult for backward compatibility — canonical home is contracts.types
__all__ = ["ObjectStoreABC", "WriteResult"]


class ObjectStoreABC(ABC):
    """ObjectStore pillar -- kernel ``file_operations`` contract.

    Linux analogue: ``struct file_operations``.
    Exception-based errors.

    Subclasses must implement the 6 abstract methods.  Streaming, batch,
    capability flags, and lifecycle have concrete defaults that work out of
    the box.
    """

    # === Identity ===

    @property
    @abstractmethod
    def name(self) -> str:
        """Backend identifier name (e.g. ``"local"``, ``"gcs"``, ``"s3"``)."""
        ...

    # === Content Operations (4 abstract) ===

    @abstractmethod
    def write_content(
        self,
        content: bytes,
        content_id: str = "",
        *,
        offset: int = 0,
        context: OperationContext | None = None,
    ) -> WriteResult:
        """Write content to storage and return a ``WriteResult``.

        Args:
            content: File content as bytes.
            content_id: Target address for the content.
                CAS backends: ignored (address = hash of content).
                PAS backends: blob path where content will be stored.
            context: Operation context (optional, for auth / cross-cutting).
            offset: Byte offset for partial write (POSIX pwrite semantics).
                0 = whole-file replace (default, backward compatible).
                >0 = splice ``content`` at offset within existing content.

        Returns:
            ``WriteResult`` with ``content_id``, ``version``, and ``size``.

        Raises:
            BackendError: If write operation fails.
        """
        ...

    @abstractmethod
    def read_content(self, content_id: str, context: OperationContext | None = None) -> bytes:
        """Read content by its opaque identifier.

        Args:
            content_id: Opaque content identifier (e.g. SHA-256 hash for CAS,
                version ID for path-based backends).
            context: Operation context (optional).

        Returns:
            File content as bytes.

        Raises:
            NexusFileNotFoundError: If content does not exist locally.
            BackendError: If read operation fails.
        """
        ...

    @abstractmethod
    def delete_content(self, content_id: str, context: OperationContext | None = None) -> None:
        """Delete content by identifier.

        Addressing-agnostic: CAS backends may defer actual deletion until
        garbage collection; PAS backends delete the blob at the given path.

        Args:
            content_id: Opaque content identifier.
            context: Operation context (optional).

        Raises:
            NexusFileNotFoundError: If content does not exist.
            BackendError: If delete operation fails.
        """
        ...

    @abstractmethod
    def get_content_size(self, content_id: str, context: OperationContext | None = None) -> int:
        """Get content size in bytes.

        Args:
            content_id: Opaque content identifier.
            context: Operation context (optional).

        Returns:
            Content size in bytes.

        Raises:
            NexusFileNotFoundError: If content does not exist.
        """
        ...

    # === Streaming (concrete defaults) ===

    @staticmethod
    def _validate_stream_chunk_size(chunk_size: int) -> None:
        if chunk_size <= 0:
            raise ValueError(f"chunk_size must be positive, got {chunk_size}")

    @classmethod
    def _validate_stream_range(cls, start: int, end: int, chunk_size: int) -> None:
        if start < 0:
            raise ValueError(f"start must be non-negative, got {start}")
        if end < start:
            raise ValueError(f"end ({end}) must be >= start ({start})")
        cls._validate_stream_chunk_size(chunk_size)

    def write_stream(
        self,
        chunks: Iterator[bytes],
        content_id: str = "",
        *,
        context: OperationContext | None = None,
    ) -> WriteResult:
        """Write content from an iterator of chunks.

        Default implementation collects all chunks into memory and
        delegates to ``write_content()``.  Backends should override for
        true streaming with incremental hashing.

        Args:
            chunks: Iterator yielding byte chunks.
            content_id: Target address (see ``write_content``).
            context: Operation context (optional).

        Returns:
            ``WriteResult`` with ``content_id``, ``version``, and ``size``.
        """
        content = b"".join(chunks)
        return self.write_content(content, content_id, context=context)

    def stream_content(
        self,
        content_id: str,
        chunk_size: int = 8192,
        context: OperationContext | None = None,
    ) -> Iterator[bytes]:
        """Stream content in chunks (generator).

        Default implementation reads the full content and yields slices.
        Backends with seekable storage should override for efficiency.

        Args:
            content_id: Opaque content identifier.
            chunk_size: Size of each chunk in bytes (default: 8 KiB).
            context: Operation context (optional).

        Yields:
            Chunks of file content.
        """
        self._validate_stream_chunk_size(chunk_size)
        content = self.read_content(content_id, context=context)
        for i in range(0, len(content), chunk_size):
            yield content[i : i + chunk_size]

    def stream_range(
        self,
        content_id: str,
        start: int,
        end: int,
        chunk_size: int = 8192,
        context: OperationContext | None = None,
    ) -> Iterator[bytes]:
        """Stream a byte range ``[start, end]`` inclusive from stored content.

        Default implementation reads full content and slices.  Backends
        with seekable storage should override for efficiency.

        Args:
            content_id: Opaque content identifier.
            start: First byte position (inclusive, 0-based).
            end: Last byte position (inclusive, 0-based).
            chunk_size: Size of each yielded chunk in bytes.
            context: Operation context (optional).

        Yields:
            Chunks covering the requested range.
        """
        self._validate_stream_range(start, end, chunk_size)
        content = self.read_content(content_id, context=context)
        sliced = content[start : end + 1]
        for i in range(0, len(sliced), chunk_size):
            yield sliced[i : i + chunk_size]

    # === Directory Operations (2 abstract) ===

    @abstractmethod
    def mkdir(
        self,
        path: str,
        parents: bool = False,
        exist_ok: bool = False,
        context: OperationContext | None = None,
    ) -> None:
        """Create a directory.

        For backends without native directory support (e.g. S3), this may
        be a no-op or create marker objects.

        Args:
            path: Directory path (relative to backend root).
            parents: Create parent directories if needed (like ``mkdir -p``).
            exist_ok: Do not raise if directory already exists.
            context: Operation context (optional).

        Raises:
            BackendError: If directory creation fails.
        """
        ...

    @abstractmethod
    def rmdir(
        self,
        path: str,
        recursive: bool = False,
        context: OperationContext | None = None,
    ) -> None:
        """Remove a directory.

        Args:
            path: Directory path.
            recursive: Remove non-empty directory (like ``rm -rf``).
            context: Operation context (optional).

        Raises:
            BackendError: If directory removal fails.
        """
        ...

    # === Batch (concrete default) ===

    def batch_read_content(
        self,
        content_ids: list[str],
        context: OperationContext | None = None,
        *,
        contexts: dict[str, OperationContext] | None = None,
    ) -> dict[str, bytes | None]:
        """Read multiple content items by their identifiers.

        Default implementation calls ``read_content()`` for each id.
        Backends should override for better performance (e.g. batch RPCs).

        Unlike ``read_content()``, missing content is indicated by
        ``None`` values in the result dict rather than raising.

        Args:
            content_ids: List of opaque content identifiers.
            context: Shared operation context (fallback).
            contexts: Per-id operation contexts mapping
                ``content_id -> OperationContext``.

        Returns:
            Dict mapping ``content_id -> bytes | None``.
        """
        result: dict[str, bytes | None] = {}
        for content_id in content_ids:
            ctx = contexts.get(content_id, context) if contexts else context
            try:
                result[content_id] = self.read_content(content_id, context=ctx)
            except Exception:
                result[content_id] = None
        return result

    def batch_write_content(
        self,
        items: list[tuple[str, bytes]],
        context: OperationContext | None = None,
        *,
        contexts: dict[str, OperationContext] | None = None,
    ) -> dict[str, WriteResult | None]:
        """Write multiple content items in a single batch.

        Default implementation calls ``write_content()`` for each item.
        Backends should override for better performance (e.g. batch RPCs).

        Args:
            items: List of ``(content_id, data)`` tuples. For CAS backends
                content_id is ignored; for PAS backends it is the blob path.
            context: Shared operation context (fallback).
            contexts: Per-id operation contexts mapping
                ``content_id -> OperationContext``.

        Returns:
            Dict mapping ``content_id -> WriteResult | None``.
            ``None`` indicates a failed write for that item.
        """
        result: dict[str, WriteResult | None] = {}
        for content_id, data in items:
            ctx = contexts.get(content_id, context) if contexts else context
            try:
                result[content_id] = self.write_content(data, content_id, context=ctx)
            except Exception:
                result[content_id] = None
        return result

    def batch_delete_content(
        self,
        content_ids: list[str],
        context: OperationContext | None = None,
        *,
        contexts: dict[str, OperationContext] | None = None,
    ) -> dict[str, bool]:
        """Delete multiple content items in a single batch.

        Default implementation calls ``delete_content()`` for each id.
        Backends should override for better performance (e.g. batch RPCs).

        Args:
            content_ids: List of opaque content identifiers.
            context: Shared operation context (fallback).
            contexts: Per-id operation contexts mapping
                ``content_id -> OperationContext``.

        Returns:
            Dict mapping ``content_id -> bool`` (True = deleted, False = failed).
        """
        result: dict[str, bool] = {}
        for content_id in content_ids:
            ctx = contexts.get(content_id, context) if contexts else context
            try:
                self.delete_content(content_id, context=ctx)
                result[content_id] = True
            except Exception:
                result[content_id] = False
        return result

    # === Lifecycle ===

    def close(self) -> None:  # noqa: B027
        """Release resources.  Consistent with MetastoreABC.close() / CacheStoreABC.close()."""

File: rust/kernel/src/abc/object_store.rs
//! ObjectStore pillar — Rust kernel `file_operations` contract.
//!
//! Rust equivalent of Python `ObjectStoreABC` (one of the Four Storage Pillars).
//! Each impl composes an addressing strategy with a transport:
//!
//!   `CasLocalBackend` = CAS addressing + LocalCASTransport
//!
//! Naming convention per `backend-architecture.md`: `<Addressing><Transport>Backend`.
//!
//! Concrete impls (`CasLocalBackend`, `PathLocalBackend`,
//! `LocalConnectorBackend`, plus the connector-feature ones) sit in
//! `_backend_impls.rs` until Phase D lifts them into the parallel
//! `backends/` crate.

use std::io;

use crate::cas_engine::CASError;

/// Error type for ObjectStore operations.
#[derive(Debug)]
#[allow(dead_code)]
pub enum StorageError {
    /// Content not found.
    NotFound(String),
    /// Underlying I/O error.
    IOError(io::Error),
    /// Operation not supported by this backend.
    NotSupported(&'static str),
}

impl From<CASError> for StorageError {
    fn from(e: CASError) -> Self {
        match e {
            CASError::NotFound(s) => StorageError::NotFound(s),
            CASError::IOError(e) => StorageError::IOError(e),
        }
    }
}

/// Result of a write operation (equivalent to Python `WriteResult`).
///
/// - `content_id`: Primary addressing key (opaque).
///   CAS backends: SHA-256 hex digest.
///   PAS backends: blob path or version ID.
/// - `version`: OCC token for conflict detection.
///   CAS: same as content_id (hash is immutable).
///   PAS: cloud version_id or content hash.
/// - `size`: Content size in bytes.
#[allow(dead_code)]
pub struct WriteResult {
    pub(crate) content_id: String,
    pub(crate) version: String,
    pub(crate) size: u64,
}

// ── ExternalTransport ──────────────────────────────────────────
// Transport-layer capability for backends that can generate direct-access
// URLs (presigned/signed). Separate from ObjectStore: not all storage
// needs this, and it belongs at the transport abstraction level, not the
// content-operations level. Only S3/GCS (and future cloud backends like
// Azure, MinIO, R2) implement this.

/// Transport-layer trait for generating direct-access URLs.
///
/// Enables clients to download/upload directly from cloud storage without
/// routing bytes through Nexus (offloading I/O, reducing memory footprint).
pub trait ExternalTransport: Send + Sync {
    /// Generate a time-limited download URL for the given object key.
    ///
    /// Returns `Ok(url_string)` on success, or `Err` if the backend cannot
    /// generate a signed URL (e.g. missing credentials).
    fn generate_download_url(
        &self,
        object_key: &str,
        expires_seconds: u64,
    ) -> Result<String, StorageError>;
}

/// ObjectStore pillar — kernel `file_operations` contract.
///
/// Rust equivalent of Python `ObjectStoreABC` (one of the Four Storage Pillars).
/// CAS/PAS agnostic: `content_id` is an opaque key whose semantics are
/// backend-defined (hash for CAS, blob path for PAS).
///
/// 6 abstract methods matching the Python ABC:
///   - write_content, read_content, delete_content, get_content_size
///   - mkdir, rmdir
///
/// Streaming (write_stream, stream_content, stream_range) and batch
/// (batch_read/write/delete) have default impls in Python; they are
/// not needed in the Rust kernel hot path and can be added later.
#[allow(dead_code)]
pub trait ObjectStore: Send + Sync {
    /// Backend identifier (e.g. "local", "gcs", "s3").
    fn name(&self) -> &str;

    /// Downcast to `&CASEngine` for CAS-specific operations. Default
    /// returns `None` for non-CAS backends (PAS, external connectors).
    /// Only `CasLocalBackend` overrides. Used by the `PyKernel::cas_*`
    /// surface so Python delegators can reach the CAS API without every
    /// backend carrying CAS-shaped noise.
    #[allow(private_interfaces)]
    fn as_cas(&self) -> Option<&crate::cas_engine::CASEngine> {
        None
    }

    /// Downcast to a streaming-capable LLM backend. Default returns `None`.
    /// Only `OpenAIBackend` (and future `AnthropicBackend`) override.
    /// Consumed by `PyKernel::llm_start_streaming` — any ObjectStore that
    /// returns `Some` must implement the full SSE → DT_STREAM →
    /// `CASEngine::write_content_tracked` pipeline.
    #[cfg(feature = "connectors")]
    fn as_llm_streaming(&self) -> Option<&dyn crate::openai_streaming::LlmStreamingBackend> {
        None
    }

    /// Write content to storage and return a `WriteResult`.
    ///
    /// - `content_id`: Target address for the content.
    ///   CAS backends: when `offset == 0`, ignored (new hash = hash of content);
    ///   when `offset > 0`, the OLD content hash the partial write is applied
    ///   against (required so CAS CDC can read+splice).
    ///   PAS backends: blob path where content will be stored.
    /// - `ctx`: Operation context (carries backend_path, auth, TTL).
    /// - `offset`: POSIX `pwrite(2)` semantics.
    ///
    ///   `offset == 0` is a full-file write (truncate + write) — current
    ///   behavior for every caller that predates the partial-write wiring
    ///   (R20.10).
    ///
    ///   `offset > 0` splices `content` starting at `offset`, preserving
    ///   bytes before `offset` and after `offset + content.len()`. When
    ///   `offset > current_size`, the gap is zero-filled (POSIX sparse-
    ///   file semantics).
    ///
    ///   Every backend that accepts `offset > 0` MUST honor this contract;
    ///   backends whose transport does not support seekable / range writes
    ///   (cloud object stores like S3, GCS — their PUT replaces the entire
    ///   object) MUST return `Err(StorageError::NotSupported)` on
    ///   `offset > 0` rather than silently falling back to read-splice-
    ///   write (that fallback would violate the caller's cost expectation
    ///   of pwrite — O(content.len()) would become O(full_blob) network
    ///   I/O).
    fn write_content(
        &self,
        content: &[u8],
        content_id: &str,
        ctx: &crate::kernel::OperationContext,
        offset: u64,
    ) -> Result<WriteResult, StorageError>;

    /// Read content by opaque identifier.
    ///
    /// `content_id`: CAS=hash, PAS=version_id.
    /// `backend_path`: mount-relative path (for path-addressed backends).
    /// `ctx`: operation credential for backends that need identity/auth.
    /// CAS backends ignore backend_path and ctx.
    fn read_content(
        &self,
        content_id: &str,
        backend_path: &str,
        ctx: &crate::kernel::OperationContext,
    ) -> Result<Vec<u8>, StorageError>;

    /// Delete content by identifier.
    fn delete_content(&self, content_id: &str) -> Result<(), StorageError> {
        let _ = content_id;
        Err(StorageError::NotSupported("delete_content"))
    }

    /// Get content size in bytes.
    fn get_content_size(&self, content_id: &str) -> Result<u64, StorageError> {
        let _ = content_id;
        Err(StorageError::NotSupported("get_content_size"))
    }

    /// Create a directory.
    fn mkdir(&self, path: &str, parents: bool, exist_ok: bool) -> Result<(), StorageError> {
        let _ = (path, parents, exist_ok);
        Err(StorageError::NotSupported("mkdir"))
    }

    /// Remove a directory.
    fn rmdir(&self, path: &str, recursive: bool) -> Result<(), StorageError> {
        let _ = (path, recursive);
        Err(StorageError::NotSupported("rmdir"))
    }

    /// Delete a file by path (PAS backends — reference-mode).
    ///
    /// CAS backends return `NotSupported` (content-addressed blobs are GC'd).
    /// PAS backends delete the physical file at the given path.
    fn delete_file(&self, path: &str) -> Result<(), StorageError> {
        let _ = path;
        Err(StorageError::NotSupported("delete_file"))
    }

    /// Rename/move a file or directory (PAS backends — reference-mode).
    ///
    /// CAS backends return `NotSupported` (paths are virtual, not physical).
    /// PAS backends rename the physical file on the host filesystem.
    fn rename(&self, old_path: &str, new_path: &str) -> Result<(), StorageError> {
        let _ = (old_path, new_path);
        Err(StorageError::NotSupported("rename"))
    }

    /// Server-side copy (PAS backends — reference-mode).
    ///
    /// Copies the physical file at `src_path` to `dst_path` within the same
    /// backend storage. CAS backends return `NotSupported` — CAS copy is
    /// metadata-only (content is deduplicated by hash), handled at the kernel
    /// layer without touching the backend.
    ///
    /// Returns `WriteResult` with the destination's content_id and size.
    fn copy_file(&self, src_path: &str, dst_path: &str) -> Result<WriteResult, StorageError> {
        let _ = (src_path, dst_path);
        Err(StorageError::NotSupported("copy_file"))
    }

    /// List direct children of a directory path.
    ///
    /// Returns entry names (not full paths). Each entry is a plain filename;
    /// directories are suffixed with `/` so callers can distinguish files
    /// from directories without a follow-up stat.
    ///
    /// Default returns `NotSupported` for backends that don't have a
    /// directory concept (CAS, remote). Filesystem backends (PathLocal,
    /// LocalConnector) use `std::fs::read_dir`; API connectors (HN, CLI, X)
    /// synthesize listings from their virtual namespace.
    fn list_dir(&self, path: &str) -> Result<Vec<String>, StorageError> {
        let _ = path;
        Err(StorageError::NotSupported("list_dir"))
    }
}
